### PR TITLE
Split repo base from output paths.

### PR DESCRIPTION
**Pull request by halfflat**
Adds support for 'out-of-tree' runs of NSuite: the base directory for working and output directories (i.e. input/, output/, cache/, config/, build/ install/) can be specified by: setting the `NS_PREFIX` environment variable; using the `--prefix` option on the top level installation and run scripts; or otherwise will default to the current working directory.

* Extend command line argument handling to be consistent across scripts, permitting long options to take their arguments with or without `=`.
* Use `BASH_SOURCE` to determine path to NSuite root, rather than relying on the root being the current working directory.
* Use `ns_prefix` as the base path for all writable directories.
* Set `ns_prefix` via `--prefix` command line option on scripts, from the `NS_PREFIX` environment variable, or from the current working directory.
* Bake `ns_prefix` into generated config env scripts.
* Force use of `ns_FOO` paths in `model_setup.sh`.
* Split `ns_output_path` into `ns_validation_output` and `ns_benchmark_output` variables.
* Conditionally build compiled validation tests and tools based on configured simulators, using a `BUILDFOR` file added to the corresponding project directories.

This constitutes work towards #16.

### CODE REVIEW

File: install-local.sh [hunk 1]
```diff
@@ -1,34 +1,40 @@
...
-    echo
+    cat <<'_end_'
+Usage: install-local.sh [-e SCRIPT] [--prefix=PATH] TARGET
+
+Setup NSuite framework, build and install simulators, benchmarks,
+and validation tests.
+
+Options:
+    -e SCRIPT       Source SCRIPT before building.
+    --prefix=PATH   Use PATH as base for install and other run-time
```
Comment (bcumming) [inline install-local.sh:10]: Can we just go with `--prefix PATH`, not handle both `--prefix=X` and `--prefix X`? to make things simpler.

And maybe replace `-e SCRIPT` with `--env SCRIPT`, to keep consistent command line argument formatting?

My preference would be:
```
    --env SCRIPT
    --prefix PATH
```

File: install-local.sh [hunk 2]
```diff
@@ -1,34 +1,40 @@
...
+                    working directories.
+
+TARGET is one of:
```
Comment (bcumming) [inline install-local.sh:13]: `is one of` doesn't make it clear that you can list more than one target

File: run-bench.sh [hunk 3]
```diff
@@ -1,74 +1,103 @@
...
-    echo
-    echo "nsuite benchmark runner:"
-    echo
-    echo "run with one, and only one, of the following options:"
-    echo "   arbor  : run arbor benchmarks"
-    echo "   neuron : run neuron benchmarks"
-    echo "   coreneuron : run coreneuron benchmarks"
-    echo
+    cat <<_end_
+Usage: run-bench.sh [--prefix=PATH] [--model MODEL] [--config CONFIG] SIMULATOR
```
Comment (bcumming) [inline run-bench.sh]: Again, the formatting of the new argument is inconsistent with the existing arguments.

Let's just pick one approach and stick with it (I prefer without the `=`).

File: scripts/environment.sh [hunk 4]
```diff
@@ -1,21 +1,25 @@
+# Set the base paths to working directories.
+# Variables defined here use the prefix ns_
+set_working_paths() {
+    if [ -z "$ns_prefix" ]; then
+	echo "error: empty ns_prefix"
```
Comment (bcumming) [inline scripts/environment.sh:5]: You can use the `err` helper function here to make for consistent error printing.

**Review by bcumming — CHANGES_REQUESTED** [review]
Nice update.

Now we can have multiple "installations" in the same path, which might also be useful.

I would prefer to use a single style for passing command line args to the install and run-bench scripts.

File: install-local.sh (hunk 1 above)
Comment (halfflat) [reply install-local.sh:10]: I've changed `-e` to `--env`. Per slack discussion, I've left both option styles (--long=FOO and --long FOO) for consistency with other software, and used '=' in the help text, following what GNU tools seem to generally do.

File: install-local.sh (hunk 2 above)
Comment (halfflat) [reply install-local.sh:13]: I've changed the first line of usage to:
```
Usage: install-local.sh [--env=SCRIPT] [--prefix=PATH] TARGET [TARGET...]
```

File: run-bench.sh (hunk 3 above)
Comment (halfflat) [reply run-bench.sh]: Fixed (currently with '=', but still happy to go the other way).

File: scripts/environment.sh (hunk 4 above)
Comment (halfflat) [reply scripts/environment.sh:5]: This is more of an internal error at this point: it 'shouldn't happen' but we should definitely stop if it does. Using `err` here would require sourcing `util.sh` from within the generated config/env scripts, which I'd rather avoid.

**Review by bcumming — APPROVED** [review]

**Comment by bcumming** [comment]
Adds support for 'out-of-tree' runs of NSuite: the base directory for working and output directories (i.e. input/, output/, cache/, config/, build/ install/) can be specified by: setting the `NS_PREFIX` environment variable; using the `--prefix` option on the top level installation and run scripts; or otherwise will default to the current working directory.

* Extend command line argument handling to be consistent across scripts, permitting long options to take their arguments with or without `=`.
* Use `BASH_SOURCE` to determine path to NSuite root, rather than relying on the root being the current working directory.
* Use `ns_prefix` as the base path for all writable directories.
* Set `ns_prefix` via `--prefix` command line option on scripts, from the `NS_PREFIX` environment variable, or from the current working directory.
* Bake `ns_prefix` into generated config env scripts.
* Force use of `ns_FOO` paths in `model_setup.sh`.
* Split `ns_output_path` into `ns_validation_output` and `ns_benchmark_output` variables.
* Conditionally build compiled validation tests and tools based on configured simulators, using a `BUILDFOR` file added to the corresponding project directories.

**Pull request merged**